### PR TITLE
Serialization of DateTimeImmutable

### PR DESCRIPTION
**Pull request by curzio-della-santa**

### CODE REVIEW

File: src/Detail/Normalization/JMSSerializer/Handler/DateHandler.php [hunk 1]
```diff
@@ -41,30 +46,65 @@ public function __construct($defaultFormat = DateTime::ISO8601, $defaultTimezone
...
-                    'type' => $type,
-                    'format' => $format,
-                    'method' => 'serialize' . $type,
-                );
-            }
+            $methods[] = array(
+                'direction' => GraphNavigator::DIRECTION_SERIALIZATION,
+                'type' => 'DateTimeImmutable',
+                'format' => $format,
+                'method' => 'serializeDateTimeImmutable' ,
```
Comment (ivan-wolf) [inline src/Detail/Normalization/JMSSerializer/Handler/DateHandler.php]: Remove space before `,`

File: src/Detail/Normalization/JMSSerializer/Handler/DateHandler.php [hunk 2]
```diff
@@ -41,30 +46,65 @@ public function __construct($defaultFormat = DateTime::ISO8601, $defaultTimezone
...
     /**
+     * @param VisitorInterface $visitor
+     * @param \DateTimeImmutable $date
+     * @param array $type
+     * @param Context $context
+     * @return string
+     */
+    public function serializeDateTimeImmutable(
+        VisitorInterface $visitor,
+        \DateTimeImmutable $date,
```
Comment (ivan-wolf) [inline src/Detail/Normalization/JMSSerializer/Handler/DateHandler.php]: No need for global resolution.

File: src/Detail/Normalization/JMSSerializer/Handler/DateHandler.php [hunk 3]
```diff
@@ -41,30 +46,65 @@ public function __construct($defaultFormat = DateTime::ISO8601, $defaultTimezone
...
+        array $type,
+        Context $context
+    ) {
+        return $this->serializeDateTime(
+            $visitor,
+            new \DateTime($date->format('Y-m-d H:i:s')),
```
Comment (ivan-wolf) [inline src/Detail/Normalization/JMSSerializer/Handler/DateHandler.php]: No need for global resolution.

File: src/Detail/Normalization/JMSSerializer/Handler/DateHandler.php [hunk 4]
```diff
@@ -103,4 +170,14 @@ protected function createDateTime($data, array $type)
 
         return $datetime;
     }
+
+    /**
+     * @param DateTime $date
+     * @return DateTimeImmutable
+     */
+    private function crateImmutableDateTimeFromMutable(DateTime $date)
```
Comment (ivan-wolf) [inline src/Detail/Normalization/JMSSerializer/Handler/DateHandler.php]: `create...`, not `crate`.

File: src/Detail/Normalization/JMSSerializer/Handler/DateHandler.php [hunk 5]
```diff
@@ -103,4 +170,14 @@ protected function createDateTime($data, array $type)
...
+    {
+        // Note DateTimeImmutable::createFromImmutable($date) available as of PHP 5.6
```
Comment (ivan-wolf) [inline src/Detail/Normalization/JMSSerializer/Handler/DateHandler.php]: Then use it if it's available.

File: src/Detail/Normalization/JMSSerializer/Handler/DateImmutableHandler.php [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    /**
+     * @param string $defaultFormat
+     * @param string $defaultTimezone
+     */
+    public function __construct($defaultFormat = DateTime::ISO8601, $defaultTimezone = 'UTC')
+    {
+        $this->format = $defaultFormat;
+        $this->timezone = new DateTimeZone($defaultTimezone);
+
+        parent::__construct($defaultFormat, $defaultTimezone);
```
Comment (ghost) [inline src/Detail/Normalization/JMSSerializer/Handler/DateImmutableHandler.php:32]: Variables are set and passed to parent constructor? Seems wrong.

File: src/Detail/Normalization/JMSSerializer/Handler/DateImmutableHandler.php [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    public function serializeDateTimeImmutable(
+        VisitorInterface $visitor,
+        DateTimeImmutable $date,
+        array $type,
+        Context $context
+    ) {
+        return $this->serializeDateTime(
+            $visitor,
+            new DateTime($date->format('Y-m-d H:i:s')),
```
Comment (ghost) [inline src/Detail/Normalization/JMSSerializer/Handler/DateImmutableHandler.php]: Shouldn't the timezone be set?

File: src/Detail/Normalization/JMSSerializer/Handler/DateImmutableHandler.php [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+     * @return DateTimeImmutable|null
+     */
+    public function deserializeDateTimeImmutableFromPhp(PhpDeserializationVisitor $visitor, $data, array $type)
+    {
+        return $this->createImmutableDateTimeFromMutable(
+            $this->deserializeDateTimeFromPhp($visitor, $data, $type)
+        );
+    }
+
+    public function deserializeDateTimeImmutableFromXml(XmlDeserializationVisitor $visitor, $data, array $type)
```
Comment (ghost) [inline src/Detail/Normalization/JMSSerializer/Handler/DateImmutableHandler.php:101]: DocBlock?

File: src/Detail/Normalization/JMSSerializer/Handler/DateImmutableHandler.php [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        return $this->createImmutableDateTimeFromMutable(
+            $this->deserializeDateTimeFromXml($visitor, $data, $type)
+        );
+    }
+
+    public function deserializeDateTimeImmutableFromJson(JsonDeserializationVisitor $visitor, $data, array $type)
```
Comment (ghost) [inline src/Detail/Normalization/JMSSerializer/Handler/DateImmutableHandler.php:114]: DocBlock?

File: src/Detail/Normalization/JMSSerializer/Handler/DateImmutableHandler.php (hunk 6 above)
Comment (curzio-della-santa) [inline src/Detail/Normalization/JMSSerializer/Handler/DateImmutableHandler.php:32]: Parent defines those variable as private, therefore are not available in local context

**Pull request merged**